common/src/main/java/com/revenuecat/purchases/common/HTTPClient.kt

```diff
@@ -101,9 +114,31 @@ class HTTPClient(
             connection.disconnect()
         }
 
-        val responseBody = payload?.let { JSONObject(it) } ?: throw IOException("Network call payload is null.")
         log(LogIntent.DEBUG, NetworkStrings.API_REQUEST_COMPLETED.format(connection.requestMethod, path, responseCode))
-        return Result(responseCode, responseBody)
+        if (payload == null) {
+            throw IOException("Network call payload is null.")
+        }
+        val result = HTTPResult(responseCode, payload)
+        val processedResponse = eTagManager.processResponse(httpRequest, connection, result)
+        return processedResponse ?: performRequest(path, body, authenticationHeaders, refreshETag = true)
```

Comment:
if the server is returning a 5xx with an empty body, there's a chance that we're going to spam it, right? 
`eTagManager` will never store the response, and it has an empty body, so we would just hit it again and again?
perhaps an easy contingency there would be that if we already had `refreshETag == true` when we get to this line, we don't retry

Reply:
also, the case where we make another request is only if something went somewhat wrong, right? 
i.e.: it's a case where the server expects us to have a response with an eTag, which we sent, but for whatever reason we don't actually have the cached response in store. 

So in this case, I think we should make it more obvious that we're seeing something odd and add a log

Reply:
Oh I see what you mean. In the case of 5xx, I believe we would store the empty result. I am going to double check and test and I'll let you know. Also, agreed to the logs. I can probably add more logs in general

Reply:
I rechecked this and I don't think this will happen. `processedResponse` is a `HTTPResult` and not only the payload, so processResponse will be an `HTTPResult` with an empty payload, so we won't retry.

Anyhow, I have refactored this part. I will also add the check to protect against calling with refresh over and over (just in case)